server/rest/src/main/java/org/infinispan/rest/server/InfinispanResponse.java

```diff
@@ -0,0 +1,153 @@
+package org.infinispan.rest.server;
+
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Date;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import org.infinispan.rest.server.operations.mediatypes.MediaType;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.http.DefaultFullHttpResponse;
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpHeaderValues;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpUtil;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.codec.http2.HttpConversionUtil;
+
+public class InfinispanResponse {
```

Comment:
> The question is, do we really need a wrapper for NettyHttpResponse? :)

It separates REST-related code from Netty. Yes, we need it.

>In any case, I'd remove the optionals, because escape analysis is not reliable enough. OTOH it requires the lambdas to be inlined (objects passed as arguments to other methods are allocated on heap), but then it's automatically disabled if the method gets too big because of inlining: http://vanillajava.blogspot.ro/2015/01/java-lambdas-and-low-latency.html. So you might get 0 allocations when you test in JMH, but then lots of allocations if toNettyHttpResponse gets inlined into a big method.

I'm glad you brought it up. I'm a big fan of `Optional`s. They simplify the code by accepting lambdas in `ifPresent(lambda_code)` and they allow chaining `optional.map().map().orThrow()`. They are really useful in `InfinispanResponse#toNettyHttpResponse` method. It is also worth to mention that we are getting a hard lesson from the reality resulting in [nearly 80 NullPointerException bugs in last 2 years](https://issues.jboss.org/issues/?jql=project%20%3D%20ISPN%20AND%20issuetype%20%3D%20Bug%20AND%20text%20~%20%22NullPointerException%22%20AND%20created%20%3E%3D%202015-04-27%20AND%20created%20%3C%3D%202017-04-27). @rvansa also asked about possibility to use it in our APIs. Now we are looking at a very concrete code example, which makes it perfect for discussion.

So far my benchmarks ([based on this with emitting JFR files and analysis via JMC](https://github.com/infinispan/infinispan/pull/5094/files#diff-014cb06f820646b99c23c65a47e0c32eR26)) showed no evidence of hot spots in any `Optional` method (nor any other caused by bad inlining). However you are absolutely correct, the JMH test checks REST interface in isolation. In order to test it properly I created [ISPN-7767](https://issues.jboss.org/browse/ISPN-7767). Unfortunately it needs to be evaluated after I update all the wiring code in the server. I plan to use similar testing procedure, start the server with JFR recording, bang it on REST interface and then check the results. Do you have any hints what I should look for?

Please also note that the blog post you shared was created in January 2015 (so probably based on testing done in Q4 2014). It has been quite a long time and I bet OpenJDK engineers implemented a lot of optimizations since then.